### PR TITLE
MinMax Index Supports Nullable DataType

### LINKED ISSUE #4787: MinMax Index Supports Nullable DataType.
## Enhancement
Currently, MinMax Index does not support `Nullable` datatype. We can extract the nested data type from `DataTypeNullable` and reuse existing MinMax index code.

### PR DESCRIPTION
**Pull request by hehechen**
Signed-off-by: hehechen <awd123456sss@gmail.com>

### What problem does this PR solve?
Currently, MinMax Index does not support Nullable datatype. 

Issue Number: close #4787

### What is changed and how it works?
We can extract the nested data type from DataTypeNullable and reuse existing MinMax index code.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/DataTypes/DataTypeNullable.h [hunk 1]
```diff
@@ -29,6 +29,7 @@ class DataTypeNullable final : public IDataType
     DataTypeNullable(const DataTypePtr & nested_data_type_);
     std::string getName() const override { return "Nullable(" + nested_data_type->getName() + ")"; }
     const char * getFamilyName() const override { return "Nullable"; }
+    DataTypePtr getNestedDataType() const override { return nested_data_type; }
```
Comment (fuzhe1989) [inline dbms/src/DataTypes/DataTypeNullable.h:32]: already has `getNestedType` at below.

File: dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp [hunk 2]
```diff
@@ -76,6 +76,10 @@ DMFileWriter::DMFileWriter(const DMFilePtr & dmfile_,
         /// for handle column always generate index
         bool do_index = cd.id == EXTRA_HANDLE_COLUMN_ID || cd.type->isInteger() || cd.type->isDateOrDateTime();
 
+        if (cd.type->isNullable() && (cd.type->getNestedDataType()->isInteger() || (cd.type->getNestedDataType()->isDateOrDateTime())))
+        {
+            do_index = true;
+        }
```
Comment (fuzhe1989) [inline dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp:82]:
```suggestion
        auto type = removeNullable(cd.type);
        bool do_index = cd.id == EXTRA_HANDLE_COLUMN_ID || type->isInteger() || type->isDateOrDateTime();
```

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 3]
```diff
@@ -82,7 +90,7 @@ void MinMaxIndex::addPack(const IColumn & column, const ColumnVector<UInt8> * de
         }
     }
 
-    const IColumn & updated_column = *column_ptr;
+    const IColumn & updated_column = column;
```
Comment (fuzhe1989) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:93]: useless? you could use `column` directly since it won't be updated.

File: dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp [hunk 4]
```diff
@@ -74,8 +74,8 @@ DMFileWriter::DMFileWriter(const DMFilePtr & dmfile_,
         // TODO: currently we only generate index for Integers, Date, DateTime types, and this should be configurable by user.
         // TODO: If column type is nullable, we won't generate index for it
```
Comment (JinheLin) [inline dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp:75]: Update these comments.

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 5]
```diff
@@ -158,6 +155,61 @@ std::pair<UInt64, UInt64> MinMaxIndex::getUInt64MinMax(size_t pack_index)
     return {minmaxes->get64(pack_index * 2), minmaxes->get64(pack_index * 2 + 1)};
 }
 
+RSResult MinMaxIndex::checkNullableEqual(size_t pack_index, const Field & value, const DataTypePtr & type)
+{
+    const ColumnNullable & column_nullable = static_cast<const ColumnNullable &>(*minmaxes);
+
+    const auto * raw_type = type.get();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:162]: Add an assert for !type->isNullable

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 6]
```diff
@@ -215,6 +271,62 @@ RSResult MinMaxIndex::checkEqual(size_t pack_index, const Field & value, const D
     }
     return RSResult::Some;
 }
+
+RSResult MinMaxIndex::checkNullableGreater(size_t pack_index, const Field & value, const DataTypePtr & type)
+{
+    const ColumnNullable & column_nullable = static_cast<const ColumnNullable &>(*minmaxes);
+    const auto * raw_type = type.get();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:278]: ditto

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 7]
```diff
@@ -272,6 +388,62 @@ RSResult MinMaxIndex::checkGreater(size_t pack_index, const Field & value, const
     }
     return RSResult::Some;
 }
+
+RSResult MinMaxIndex::checkNullableGreaterEqual(size_t pack_index, const Field & value, const DataTypePtr & type)
+{
+    const ColumnNullable & column_nullable = static_cast<const ColumnNullable &>(*minmaxes);
+
+    const auto * raw_type = type.get();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:396]: ditto

File: dbms/src/Storages/DeltaMerge/tests/gtest_dm_minmax_index.cpp [hunk 8]
```diff
@@ -236,6 +228,73 @@ try
...
+    ASSERT_EQ(true, checkMatch(case_name, *context, "Nullable(MyDateTime)", "2020-09-27", createGreaterEqual(attr("Nullable(MyDateTime)"), parseMyDateTime("2020-09-27"), 0)));
+    ASSERT_EQ(false, checkMatch(case_name, *context, "Nullable(MyDateTime)", "2020-09-27", createGreaterEqual(attr("Nullable(MyDateTime)"), parseMyDateTime("2020-09-28"), 0)));
+    ASSERT_EQ(true, checkMatch(case_name, *context, "Nullable(MyDateTime)", "2020-09-27", createLess(attr("Nullable(MyDateTime)"), parseMyDateTime("2020-09-28"), 0)));
+    ASSERT_EQ(false, checkMatch(case_name, *context, "Nullable(MyDateTime)", "2020-09-27", createLess(attr("Nullable(MyDateTime)"), parseMyDateTime("2020-09-27"), 0)));
+    ASSERT_EQ(true, checkMatch(case_name, *context, "Nullable(MyDateTime)", "2020-09-27", createLessEqual(attr("Nullable(MyDateTime)"), parseMyDateTime("2020-09-27"), 0)));
+    ASSERT_EQ(false, checkMatch(case_name, *context, "Nullable(MyDateTime)", "2020-09-27", createLessEqual(attr("Nullable(MyDateTime)"), parseMyDateTime("2020-09-26"), 0)));
+
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_dm_minmax_index.cpp:287]: Add test case for all values in a pack is `NULL`

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp (hunk 5 above)
Comment (hehechen) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:162]: > Add an assert for !type->isNullable

Caller has checked the type with `typeid_cast<const DataTypeNullable *>` and extracted the nested type with `removeNullable(type)`.

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp (hunk 7 above)
Comment (hehechen) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:396]: > ditto

Caller has checked the type with typeid_cast<const DataTypeNullable *> and extracted the nested type with removeNullable(type).

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp [hunk 9]
```diff
@@ -158,6 +155,61 @@ std::pair<UInt64, UInt64> MinMaxIndex::getUInt64MinMax(size_t pack_index)
...
 
+RSResult MinMaxIndex::checkNullableEqual(size_t pack_index, const Field & value, const DataTypePtr & type)
+{
+    const ColumnNullable & column_nullable = static_cast<const ColumnNullable &>(*minmaxes);
+
+    const auto * raw_type = type.get();
+#define DISPATCH(TYPE)                                                                         \
+    if (typeid_cast<const DataType##TYPE *>(raw_type))                                         \
+    {                                                                                          \
+        auto & minmaxes_data = toColumnVectorData<TYPE>(column_nullable.getNestedColumnPtr()); \
```
[the remark this answers is not in the document]
Comment (hehechen) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:166]: > For a nullable column, if `column[i]` is null, then the value of `nested_column[i]` is meaningless and maybe just some random value. So we can't just get the nested column and use the value at the specified index to do the check.

If MinMax Indexes have null value, the `has_null_marks` will be true, and skip the MinMax Indexes check.

[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:166]: Seems we need to add some comments before line 215

[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.cpp:166]: Got it.

**Review by lidezhu — APPROVED** [review]
LGTM

**Review by JaySon-Huang — APPROVED** [review]
LGTM
